homeassistant/components/api/__init__.py

```diff
@@ -385,6 +385,44 @@ async def post(
             )
 
         context = self.context(request)
+
+        if "return_response" in request.query or (
+            data is not None
+            and isinstance(data, dict)
+            and data.pop("return_response", False)
+        ):
+            if (
+                hass.services.supports_response(domain, service)
+                == ha.SupportsResponse.NONE
+            ):
+                return self.json_message(
+                    "Service does not support responses. Remove return_response from request.",
+                    HTTPStatus.BAD_REQUEST,
+                )
+
+            try:
+                # shield the service call from cancellation on connection drop
+                response = await shield(
+                    hass.services.async_call(
+                        domain,
+                        service,
+                        data,  # type: ignore[arg-type]
+                        blocking=True,
+                        context=context,
+                        return_response=True,
+                    )
+                )
+            except (vol.Invalid, ServiceNotFound) as ex:
+                raise HTTPBadRequest from ex
+
+            return self.json(response)
```

Comment:
We should return a dict with the response and changed states:
```suggestion
            return self.json({"changed_states": changed_states, "service_response": response})
```

Reply:
Cool, that's what I had suggested in the latest version of the architecture proposal.

Reply:
Done, please take a look and resolve if it's all good.